File: swift/Solutions/Day08.swift
//
//  Day08.swift
//  advent_of_code_2023
//
//  Created by Kushantha Attanayake on 12/7/23.
//

import Foundation

struct Day08: Day {
    func part1Solution(input: [String]) -> String {
        let instructions = input[0]
        var nodes = [String: [String]]()
        input[1...].forEach { line in
            let lineParts = line.split(separator: "=")
            nodes[String(lineParts[0]).trimmingCharacters(in: .whitespaces)] = lineParts[1].split(separator: ",").map{
                $0.trimmingCharacters(in: CharacterSet(charactersIn: "() ") )
            }
        }
        var curNode = "AAA"
        var curIndex = 0
        var answer = 0
        
        while(curNode != "ZZZ"){
            let direction = instructions[curIndex]
            let newIndex: Int
            if (direction == "L") { newIndex = 0} else { newIndex = 1}
            curNode = nodes[curNode]![newIndex]
            answer += 1
            curIndex = (curIndex + 1) % instructions.count
        }
        
        return "\(answer)"
    }
    
    func part2Solution(input: [String]) -> String {
        let instructions = input[0]
        var nodes = [String: [String]]()
        input[1...].forEach { line in
            let lineParts = line.split(separator: "=")
            nodes[String(lineParts[0]).trimmingCharacters(in: .whitespaces)] = lineParts[1].split(separator: ",").map{
                $0.trimmingCharacters(in: CharacterSet(charactersIn: "() ") )
            }
        }
        
        var curNodes = nodes.keys.filter { $0.last == "A" }
        var curIndex = 0
        var iteration = 0
        var LCMs = [Int](repeating: -1, count: curNodes.count)
        while(!LCMs.allSatisfy { $0 != -1}){
            let direction = instructions[curIndex]
            let newIndex: Int
            if (direction == "L") { newIndex = 0} else { newIndex = 1}
            for (idx, curNode) in curNodes.enumerated() {
                curNodes[idx] = nodes[curNode]![newIndex]
                if(curNodes[idx].last == "Z" && LCMs[idx] == -1){
                    LCMs[idx] = iteration + 1
                }
            }
            iteration += 1
            curIndex = (curIndex + 1) % instructions.count
        }
        
        var accumulator = 1
        for value in LCMs {
            accumulator = lcm(accumulator, value)
        }
        
        return "\(accumulator)"
    }
    
    
}

File: kotlin/src/main/kotlin/solutions/Day08.kt
package solutions

import helpers.lowestCommonMultiple
import kotlin.math.max
import kotlin.math.min

class Day08: Day {
    override fun part1Solution(input: List<String>): String {
        val instructions = input[0]
        val nodes = mutableMapOf<String, List<String>>()
        for(lineIndex in 2..<input.size){
            val lineParts = input[lineIndex].split("=")
            nodes[lineParts[0].trim()] = lineParts[1].split(",").map{
                it.trim('(',')',' ')
            }
        }
        var curNode = "AAA"
        var curIndex = 0
        var answer = 0

        while(curNode != "ZZZ"){
            val direction = instructions[curIndex]
            val newIndex = if (direction == 'L') 0 else 1
            curNode = nodes[curNode]!![newIndex]
            answer += 1
            curIndex = (curIndex + 1) % instructions.length
        }

        return answer.toString()
    }

    override fun part2Solution(input: List<String>): String {
        val instructions = input[0]
        val nodes = mutableMapOf<String, List<String>>()
        for(lineIndex in 2..<input.size){
            val lineParts = input[lineIndex].split("=")
            nodes[lineParts[0].trim()] = lineParts[1].split(",").map{
                it.trim('(',')',' ')
            }
        }
        val curNodes = nodes.keys.filter { it.last() == 'A' }.toMutableList()
        var curIndex = 0
        var iterations = 0
        val LCMs = IntArray(curNodes.size) { -1 }
        while(!LCMs.all { it != -1 }){
            val direction = instructions[curIndex]
            val newIndex = if (direction == 'L') 0 else 1
            curNodes.forEachIndexed { nodeIndex, curNode ->
                curNodes[nodeIndex] = nodes[curNode]!![newIndex]
                if (curNodes[nodeIndex].last() == 'Z' && LCMs[nodeIndex] == -1){
                    LCMs[nodeIndex] = iterations + 1
                }
            }
            iterations += 1
            curIndex = (curIndex + 1) % instructions.length
        }

        var accumulator = 1L
        for (lcm in LCMs){
            accumulator = lowestCommonMultiple(accumulator, lcm.toLong())
        }

        return accumulator.toString()
    }


}